src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketDuplicationTests.cs

```diff
@@ -317,44 +317,56 @@ public async Task DuplicateAndClose_TcpServerHandler(AddressFamily addressFamily
+                ipcServerListener.Listen(1);
+                string ipcPortString = ((IPEndPoint)ipcServerListener.LocalEndPoint).Port.ToString(CultureInfo.InvariantCulture);
 
                 if (sameProcess)
                 {
-                    Task handlerCode = Task.Run(() => HandlerServerCode(_ipcPipeName));
-                    await RunCommonHostLogic(Environment.ProcessId);
+                    Task handlerCode = Task.Run(() => HandlerServerCode(ipcPortString));
+                    RunCommonHostLogic(Environment.ProcessId);
                     await handlerCode;
                 }
                 else
                 {
-                    RemoteInvokeHandle hServerProc = RemoteExecutor.Invoke(HandlerServerCode, _ipcPipeName);
-                    await RunCommonHostLogic(hServerProc.Process.Id);
-                    await hServerProc.DisposeAsync();
+                    RemoteInvokeHandle hServerProc = RemoteExecutor.Invoke(HandlerServerCode, ipcPortString);
+
+                    // Since RunCommonHostLogic can throw, we need to make sure the server process is disposed
+                    try
+                    {
+                        RunCommonHostLogic(hServerProc.Process.Id);
+                    }
+                    finally
+                    {
+                        await hServerProc.DisposeAsync();
+                    }
                 }
 
-                async Task RunCommonHostLogic(int processId)
+                void RunCommonHostLogic(int processId)
                 {
-                    await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
-                    
-                    pipeServerStream.WaitForConnection();
+                    using Socket ipcServer = ipcServerListener.Accept();
 
                     // Duplicate the socket:
                     SocketInformation socketInfo = handlerOriginal.DuplicateAndClose(processId);
-                    WriteSocketInfo(pipeServerStream, socketInfo);
+                    // Send socketInfo to server code
+                    using var ipcStream = new NetworkStream(ipcServer, false);
+                    WriteSocketInfo(new NetworkStream(ipcServer), socketInfo);
 
                     // Send client data:
                     client.Send(TestBytes);
                 }
 
-                static async Task<int> HandlerServerCode(string ipcPipeName)
+                static async Task<int> HandlerServerCode(string ipcPortString)
                 {
-                    await using NamedPipeClientStream pipeClientStream =
-                        new NamedPipeClientStream(".", ipcPipeName, PipeDirection.In);
-                    pipeClientStream.Connect();
+                    int ipcPort = int.Parse(ipcPortString, CultureInfo.InvariantCulture);
+                    using Socket ipcClient = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
```

Comment:
nit: I would also add `ipcClient.NoDelay = true;` (same for the other side)

Reply:
Great hint. Done.